File: src/javascript/algorithms/linkedlist/7.findMiddleNode/findMiddleNode.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class LinkedList {
    constructor(value) {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    push(value) {
        const newNode = new Node(value);
        if(!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.length++;
        return this;
    }

    findMiddleNode() {
        let slow = this.head;
        let fast = this.head;
        while(fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }

    findMiddleNodeWithLength() {
        let temp = this.head;
        for(let i = 0; i < Math.floor(this.length/2); i++) {
            temp = temp.next;
        }
        return temp;
    }
}

const myLinkedList1 = new LinkedList();
myLinkedList1.push(1);
myLinkedList1.push(2);
myLinkedList1.push(3);
myLinkedList1.push(4);
myLinkedList1.push(5);

const myLinkedList2 = new LinkedList();
myLinkedList2.push(1);
myLinkedList2.push(2);
myLinkedList2.push(3);
myLinkedList2.push(4);
myLinkedList2.push(5);
myLinkedList2.push(6);

console.log(myLinkedList1.findMiddleNode());
console.log(myLinkedList1.findMiddleNodeWithLength());

console.log(myLinkedList2.findMiddleNode());
console.log(myLinkedList2.findMiddleNodeWithLength());

File: src/python/algorithms/linkedlist/7.findMiddleNode/find_middle_node.py
"""
Find Middle Node of Linked List

Given the head of a singly linked list, return the middle node.
If there are two middle nodes, return the second one.

Time Complexity: O(n)
Space Complexity: O(1)
"""


class ListNode:
    def __init__(self, value=0):
        self.value = value
        self.next = None


def find_middle_node(head):
    """
    Two pointer approach (slow and fast).
    TC: O(n), SC: O(1)
    """
    if not head:
        return None
    
    slow = head
    fast = head
    
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
    
    return slow


def print_list(head):
    """Helper to print linked list."""
    result = []
    while head:
        result.append(str(head.value))
        head = head.next
    return " -> ".join(result)


# Test cases
if __name__ == "__main__":
    # Create list: 1 -> 2 -> 3 -> 4 -> 5
    head1 = ListNode(1)
    head1.next = ListNode(2)
    head1.next.next = ListNode(3)
    head1.next.next.next = ListNode(4)
    head1.next.next.next.next = ListNode(5)
    
    middle = find_middle_node(head1)
    print(f"List: {print_list(head1)}")
    print(f"Middle node value: {middle.value}")  # 3
    
    # Create list: 1 -> 2 -> 3 -> 4 -> 5 -> 6
    head2 = ListNode(1)
    head2.next = ListNode(2)
    head2.next.next = ListNode(3)
    head2.next.next.next = ListNode(4)
    head2.next.next.next.next = ListNode(5)
    head2.next.next.next.next.next = ListNode(6)
    
    middle2 = find_middle_node(head2)
    print(f"List: {print_list(head2)}")
    print(f"Middle node value: {middle2.value}")  # 4
